dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+{
+    if (!ppFile)
+        return E_POINTER;
+
+    LPCWSTR Extension = wcsrchr(pDeletedFile->FileNameW, '.');
+    WCHAR szFullName[MAX_PATH];
+    wsprintfW(szFullName, L"%s\\D%c%lu%s", Folder, pDeletedFile->dwDriveNumber + 'a', pDeletedFile->dwRecordUniqueId, Extension);
+    if (GetFileAttributesW(szFullName) == INVALID_FILE_ATTRIBUTES)
+        return E_FAIL;
+
+    LPWSTR pszFullName;
+    HRESULT hr = SHStrDup(szFullName, &pszFullName);
+    if (FAILED(hr))
+        return hr;
+
+    RecycleBin5File *pThis = new RecycleBin5File(prb, Folder, pDeletedFile, pszFullName);
+    if (!pThis)
+        return E_OUTOFMEMORY;
+
+    *ppFile = static_cast<IRecycleBinFile *>(pThis);
+    return S_OK;
+}
+
+class RecycleBin5Enum : public IRecycleBinEnumList
+{
+public:
+    RecycleBin5Enum(
+        IN IRecycleBin5 *prb,
+        IN HANDLE hInfo,
+        IN OUT INFO2_HEADER *pInfo,
+        IN OUT LPWSTR pszPrefix);
+    virtual ~RecycleBin5Enum();
+
+    /* IUnknown methods */
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject) override;
+    STDMETHODIMP_(ULONG) AddRef() override;
+    STDMETHODIMP_(ULONG) Release() override;
+
+    /* IRecycleBinEnumList methods */
+    STDMETHODIMP Next(DWORD celt, IRecycleBinFile **rgelt, DWORD *pceltFetched) override;
+    STDMETHODIMP Skip(DWORD celt) override;
+    STDMETHODIMP Reset() override;
+
+protected:
+    LONG m_ref;
+    IRecycleBin5 *m_recycleBin;
+    HANDLE m_hInfo;
+    INFO2_HEADER *m_pInfo;
+    DWORD m_dwCurrent;
+    LPWSTR m_pszPrefix;
+};
+
+STDMETHODIMP RecycleBin5Enum::QueryInterface(REFIID riid, void **ppvObject)
+{
+    TRACE("(%p, %s, %p)\n", this, debugstr_guid(&riid), ppvObject);
+
+    if (!ppvObject)
+        return E_POINTER;
+
+    if (IsEqualIID(riid, IID_IUnknown) || IsEqualIID(riid, IID_IRecycleBinEnumList))
+        *ppvObject = static_cast<IRecycleBinEnumList *>(this);
+    else
+    {
+        *ppvObject = NULL;
+        return E_NOINTERFACE;
+    }
+
+    AddRef();
+    return S_OK;
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::AddRef()
+{
+    TRACE("(%p)\n", this);
+    return InterlockedIncrement(&m_ref);
+}
+
+RecycleBin5Enum::~RecycleBin5Enum()
+{
+    TRACE("(%p)\n", this);
+
+    m_recycleBin->OnClosing(this);
+    UnmapViewOfFile(m_pInfo);
+    m_recycleBin->Release();
+    SHFree(m_pszPrefix);
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5Enum::Release()
+{
+    TRACE("(%p)\n", this);
+
+    ULONG refCount = InterlockedDecrement(&m_ref);
+    if (refCount == 0)
+        delete this;
+    return refCount;
+}
+
+STDMETHODIMP RecycleBin5Enum::Next(DWORD celt, IRecycleBinFile **rgelt, DWORD *pceltFetched)
+{
+    HRESULT hr;
+
+    TRACE("(%p, %u, %p, %p)\n", this, celt, rgelt, pceltFetched);
+
+    if (!rgelt)
+        return E_POINTER;
+    if (!pceltFetched && celt > 1)
+        return E_INVALIDARG;
+
+    ULARGE_INTEGER FileSize;
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
+    if (FileSize.u.LowPart == 0)
```

Comment:
There's a bug here, because according to MSDN the correct check to see whether there has been an error in the GetFileSize call would have to read something like:
```c
        if ((FileSize.u.LowPart  == INVALID_FILE_SIZE) && (GetLastError() != ERROR_SUCCESS))
```
Otherwise I think you could use GetFileSizeEx :thinking:

This change may be done in a separate PR, or at the very least in a separate commit (since the current code is what was originally present)